### DISCUSSION TITLE
Adding support for dual boot with Windows

### DISCUSSION THREAD

**Discussion by kquinsland** [Sithuk/ubuntu-server-zfsbootmenu#23] [Ideas]
Hi!  I found your script and am grateful that somebody's taken the time to build one... better than I could have!

I was wondering if you had thoughts on how best to add support for dual booting windows.

I was going to fork the script or just make local changes as needed to the partitioning, but it occurred to me that it _might_ make sense to adjust the script _slightly_ to use something like `source` to load in user specified partitioning for a more flexible approach.

Basically, the approach I was considering was changing [`partitionsFunc()`](https://github.com/Sithuk/ubuntu-server-zfsbootmenu/blob/main/ubuntu_server_encrypted_root_zfs.sh#L332) ever so slightly:

```shell
        ALT_PARTITION_CMDS=./dualboot_win_partitions.sh
        
        while IFS= read -r diskidnum;
        do
            if [[ -f "$ALT_PARTITION_CMDS" ]]; then
                echo "Using alternative partitioning from '$ALT_PARTITION_CMDS'."
                source $ALT_PARTITION_CMDS
            else

                echo "Creating partitions on disk ${diskidnum}."
                ##2.3 create bootloader partition
                sgdisk -n1:1M:+"$EFI_boot_size"M -t1:EF00 /dev/disk/by-id/"${diskidnum}"
            
                ##2.4 create swap partition 
                ##bug with swap on zfs zvol so use swap on partition:
                ##https://github.com/zfsonlinux/zfs/issues/7734
                ##hibernate needs swap at least same size as RAM
                ##hibernate only works with unencrypted installs
                sgdisk -n2:0:+"$swap_size"M -t2:"$swap_hex_code" /dev/disk/by-id/"${diskidnum}"
            
                ##2.6 Create root pool partition
                ##Unencrypted or ZFS native encryption:
                sgdisk     -n3:0:0      -t3:BF00 /dev/disk/by-id/"${diskidnum}"
            fi

        done < /tmp/diskid_check_"${pool}".txt
```

And the content of `dualboot_win_partitions.sh` would be along the lines of:

```
echo "Creating partitions on disk ${diskidnum}."

##
# We use similar partition scheme to the 'default' script but instead of letting the 3rd partition (zfs / root dataset) occupy the 
#   rest of the disk, we carve out some space at the end for windows. Specifically, windows needs - as far as I can tell - two
#   partitions in addition to the EFI/Boot partition.
##

##2.3 create bootloader partition
sgdisk --new=1:1M:+"$EFI_boot_size"M -t1:EF00 /dev/disk/by-id/"${diskidnum}"
            
##2.4 create swap partition 
sgdisk --new=2:0:+"$swap_size"M -t2:"$swap_hex_code" /dev/disk/by-id/"${diskidnum}"
            
##2.6 Create root pool partition for *nix on ZFS, leaving the last 300 gigs for windows
sgdisk     --new=3:0:-300GiB      -t3:BF00 /dev/disk/by-id/"${diskidnum}"

## Create windows primary partition
sgdisk     --new=4:0:-2GiB      -t4:0700 /dev/disk/by-id/"${diskidnum}"

## Create windows recovery partition
sgdisk     --new=4:0:0      -t4:0700 /dev/disk/by-id/"${diskidnum}"
```

Is this an acceptable approach or is there a better/different way that would play nicely with your script?

**Comment by Sithuk** [comment]
Hi kquinsland, I'm glad you've found the script helpful. Your approach to partitioning should work fine. Please post back here if you test it successfully.

Making the script dual boot windows compatible is a little outside the core aim for a remotely unlockable headless server. If you confirm the approach works, I might put a link to this thread in the FAQ section in case there are any other windows dual booters who would benefit.

Your sourcing approach for an external partition layout looks ok too. I've also avoided sourcing external files so I can keep everything in one place for ease of admin. But theres no reason it shouldn't work for you.

Please let me know how you get on.
